torch/distributed/_composable/fsdp/_fsdp_init.py

```diff
@@ -11,6 +11,61 @@
 from ._fsdp_state import _get_module_fsdp_state
 
 
+def _get_post_forward_mesh_info(
+    reshard_after_forward: Union[bool, int], mesh_info: FSDPMeshInfo
+) -> Optional[FSDPMeshInfo]:
+    shard_mesh_size = mesh_info.shard_mesh_size
+    if not isinstance(reshard_after_forward, (bool, int)):
+        raise ValueError(
+            "reshard_after_forward should be a bool or an int representing the "
+            f"group size to reshard to, not {reshard_after_forward}"
+        )
+    # NOTE: `isinstance(False, int)` returns `True`.
+    if not isinstance(reshard_after_forward, bool) and isinstance(
+        reshard_after_forward, int
+    ):
+        if (
+            reshard_after_forward < 1
+            or reshard_after_forward > shard_mesh_size
+            or shard_mesh_size % reshard_after_forward != 0
+        ):
+            raise ValueError(
+                "If passing reshard_after_forward as an int, it should be a "
+                f"factor of {shard_mesh_size}, not {reshard_after_forward}"
+            )
+        elif reshard_after_forward == 1:
+            reshard_after_forward = False
+        elif reshard_after_forward == shard_mesh_size:
+            reshard_after_forward = True
+    post_forward_mesh_info = None
+    if reshard_after_forward is True:
+        post_forward_mesh_info = mesh_info
+    elif reshard_after_forward is not False:  # int case
+        post_forward_shard_mesh_size = reshard_after_forward
+        num_post_forward_meshes = (
+            mesh_info.shard_mesh_size // post_forward_shard_mesh_size
+        )
+        shard_pg = mesh_info.mesh.get_group(mesh_info.shard_mesh_dim)
+        assert isinstance(shard_pg, dist.ProcessGroup)  # mypy
+        mesh_shard_ranks = sorted(
```

Comment:
Another detail here that we may change in follow-ups:

Here, we manually construct a 1D submesh of the 1D FSDP shard mesh. Instead, we could:
- For FSDP, view the FSDP shard mesh from 1D to 2D with shape `(shard_mesh_size // reshard_after_forward, reshard_after_forward)`
- For HSDP, view the HSDP mesh from 2D to 3D with shape `(replicate_mesh_size, shard_mesh_size // reshard_after_forward, reshard_after_forward)`

I am not sure about the upsides/downsides of doing that. The main downside would be constructing additional NCCL PGs for the replicate dimension introduced by this `reshard_after_forward`, where that replicate dimension should never be used (at least today).

Reply:
I changed it to the following to leverage existing `DeviceMesh` infra:
```
post_forward_mesh_tensor = mesh_info.mesh.mesh.view(-1, reshard_after_forward)
post_forward_mesh = DeviceMesh(
    mesh_info.mesh.device_type, post_forward_mesh_tensor
)
post_forward_mesh_info = HSDPMeshInfo(
    post_forward_mesh, shard_mesh_dim=1, replicate_mesh_dim=0
)
```
This does mean we create the extra NCCL PGs for the replicate dim though.

For HSDP case, I think we are okay to flatten the two replicate dims into the 0th dim. E.g. if we had HSDP with 4-way replication and 8-way sharding, i.e. `(4, 8)`), and we had `reshard_after_forward=2`, then instead of `(4, 4, 2)`, we flatten to `(16, 2)` since we only care about the shard dim really.